federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,956 @@
+	"k8s.io/kubernetes/federation/pkg/dnsprovider"
+	"k8s.io/kubernetes/federation/pkg/dnsprovider/rrstype"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/conversion"
+)
+
+const (
+
+	serviceHashLabelName = "service-spec-hash"
+	serviceSyncPeriod = 30 * time.Second
+	clusterSyncPeriod = 100 * time.Second
+// How long to wait before retrying the processing of a service change.
+// If this changes, the sleep in hack/jenkins/e2e.sh before downing a cluster
+// should be changed appropriately.
+	minRetryDelay = 5 * time.Second
+	maxRetryDelay = 300 * time.Second
+
+	clientRetryCount    = 5
+	clientRetryInterval = 5 * time.Second
+
+	retryable    = true
+	notRetryable = false
+
+	doNotRetry = time.Duration(0)
+
+	UserAgentName = "Federation-Service-Controller"
+	KubeAPIQPS    = 20.0
+	KubeAPIBurst  = 30
+)
+
+type cachedService struct {
+
+	//release_1_3 client returns v1 service
+	lastState *v1.Service
+	// The state as successfully applied to the load balancer
+	appliedState *v1.Service
+
+	// Ensures only one goroutine can operate on this service at any given time.
+	mu sync.Mutex
+
+	// Controls error back-off
+	lastRetryDelay time.Duration
+}
+
+type serviceCache struct {
+	mu         sync.Mutex // protects serviceMap
+	serviceMap map[string]*cachedService
+}
+
+type ServiceController struct {
+	dns                dnsprovider.Interface
+	knownClusterSet    sets.String
+	//we need two clientsets to talk to federation apiserver now, federationClient talks to cluster
+	//kubeclient talks to services.
+	federationClient   federationclientset.Interface
+	federationName     string
+	clusterClientMap   map[string]clientset.Clientset
+	balancer           cloudprovider.LoadBalancer
+	zones              []dnsprovider.Zone
+	cache              *serviceCache
+	clusterClientCache *clusterClientCache
+	// A store of services, populated by the serviceController
+	serviceStore       cache.StoreToServiceLister
+	// Watches changes to all services
+	serviceController  *framework.Controller
+	// A store of services, populated by the serviceController
+	clusterStore       federationcache.StoreToClusterLister
+	// Watches changes to all services
+	clusterController  *framework.Controller
+	eventBroadcaster   record.EventBroadcaster
+	eventRecorder      record.EventRecorder
+	//clusterLister     federationcache.StoreToClusterLister
+	// services that need to be synced
+	queue              *workqueue.Type
+}
+
+// New returns a new service controller to keep cloud provider service resources
+// (like Kubernetes Services and DNS server records for service discovery) in sync with the registry.
+
+func New(federationClient federationclientset.Interface, dns dnsprovider.Interface, federationName string) *ServiceController {
+	// TODO: quinton:
+	//                 1. cloudprovider.Interface is used for DNS configuration?
+	//                 2. need one kubeClient per cluster, or look up based on type of cluster.
```

Comment:
We will create clientset(version: release_1_3 only) for each cluster and put them into clusterClientMap, and for each cluster, we would create a informcontroller to watch the services changes.